### ISSUE TITLE
[RFC] Suggested new warp effects for Warp Theory

### ISSUE THREAD

**Issue by D-Cysteine** [GTNewHorizons/GT-New-Horizons-Modpack#8611]
### Your GTNH Discord Username

D-Cysteine

### Your Pack Version

2.1.1.3

### Your Proposal

Here's some ideas that I came up with; please give feedback and/or suggest new ones!

In roughly suggested order of increasing warp:
 * For a short period of time, flowers are spawned at the player's feet (if placeable). Maybe limit to grass blocks only; this shouldn't break anything in bases.
 * Player receives slowness, weakness, mining fatigue.
 * For a period of time, player randomly lays eggs.
 * For a period of time, player either cannot sleep, or takes twice as long to fall asleep.
 * Spawn a wisp near the player.
 * Player instantly loses half their current health (rounded down so this cannot kill the player).
 * Spawn a doppelgänger behind the player. Similar to the friendly creeper, the doppelgänger is a passive mob that follows the player. The doppelgänger looks just like the player and has the same max health, and any damage that it receives is also dealt to the player (maybe with a chat message each time this happens?).
 * Spawn some hostile endermen a moderate distance away. Maybe also give the player an ongoing effect that causes nearby endermen to become hostile and/or randomly spawns more endermen.
 * For a period of time, the player randomly gets pushed around.
 * Water is converted to ice in an area around the player. Maybe also place some snow.
 * Spawn a bunch of mushrooms in an area around the player (low chance of vishroom). Cows in range are converted to mooshrooms (need to figure out how this interacts with Thaumic Horizons infused cows).
 * _A voice whispers in your ear, "You look like you need this"._ Gives the player some litmus paper. This is a not-so-subtle hint that the player is about to reach (or maybe has reached) wither-level warp. Maybe give the player an effect that prevents receiving this warp effect again for a long time, to prevent getting free litmus paper?

I'm happy to work on implementing these myself.

For balancing, note that it looks like a random event is picked (out of those the player has enough warp for), so the more events that we add, the less likely any individual event is to be picked. See:
https://github.com/GTNewHorizons/WarpTheory/blob/master/src/main/java/shukaro/warptheory/handlers/WarpHandler.java#L228-L238

There is also a closed issue with more ideas: https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/4270

### Your Goal

According to https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/5643, we need new warp effects implemented so that we can rebalance warp.

### Your Vision

Adding new warp effects will allow us to expand the range of warp effects, pushing back destructive effects farther and allowing Thaumcraft users to be less afraid of withers blowing up their base.

### Final Checklist

- [X] I have searched this issue tracker and there is nothing similar already. Posting on a closed issue saying *I like this feature please reconsider adding it* will prompt us to investigate and reopen it once we confirm your report.
- [X] I believe there is nothing similar in the pack already, or the existing solution isn't good enough.
- [X] I understand this change request may not attract enough attention and thus not be implemented.
- [X] I understand this change request may be rejected due to other community members thinking it's inappropriate.
- [X] I believe this feature would make the pack better.

**Comment by boubou19** [comment]
That is actually a bunch of really cool additions, i personally like a lot the idea of having a doppelganger linked to your health. I also love the last one, as it's basically telling the player when it's time to care about warp, so even begineers in thaumcraft don't get trapped that easily with warp levels.

That might also let us spread way more the warp effects, and defacto push back the wither boss warp to higher levels without any issue.

**Comment by yukieiji** [comment]
It's closed now, but in the past there was some discussion about whether the following Warp effects should be added. #4270

**Comment by D-Cysteine** [comment]
Yeah, I think the doppelgänger is also my favorite of these. There's some fun trolling potential too on multiplayer.

Thanks for mentioning #4270! I think I read it a while ago (probably got the wisp idea from there actually), but didn't find it when I was creating this issue because I only searched open issues. I've added it to the first comment.

I do think that a lot of the ideas in #4270 may be quite tricky to implement, though. Swapping controls and checking if the player has removed an item of armor seem like they might be difficult to do, and affecting GT machines would require WarpTheory to depend on GT5.

**Comment by bombcar** [comment]
I like these and I also would suggest another warp effect - kill all nearby bats! Not all warps have to be bad ;). 

We may also want to revisit the warp researches and if they still apply - the way I look at it for some warp is the penalty paid to get something earlier than tech would get it (so a warped thing at HV that tech can’t do until EV) for example.

Another might be shuffling just player inventory.

**Comment by D-Cysteine** [comment]
Similar to this, one of the scrapped ideas I thought of was spawning in bats, but with wither on them so that they die after a short time (as a twist / higher-warp version of the regular bat swarm). But I decided against it because it would probably be really loud to have them all die at the same time, and the effect is arguably less annoying than the regular bat swarm.

**Comment by SteamDoge9000** [comment]
I got a better idea

This sort of goes with 1.12.2 bat mechanics

but think of this, instead of bats, the bats are magical spellbats. they are projectiles and being hit by one will give you a effect/debuff

**Comment by SteamDoge9000** [comment]
(Basically if the bat touches you it may have a chance of despawning on contact and give you a debuff

the amount of warp that you have will determine the debuff power, the duration, and if the bat that touches you will despawn (goes down in despawn chance when touching you as your warp increases) )

Example: if the minium for this effect is 40, then theres is a X chance that the bat will despawn when it touches you after it applies the debuff. but when your at 80, the chances are lower of it despawning and the effects it can give you can varry

But why stop at just thaumcraft effects, maybe some debuffs from other mods can be piled on to if (Ie, witchery effects that can be a pain) it all depends on the current pool of warp they are having

Im thinking this being done on an equation where the total warp value is pulled and then used to determine the effects that the bats can give you 

on another note, because witchery is in this pack, the bats colors can be altered with a simple potion effect that can be applied onto them when they spawn and diffrent colors can potentaly represent different debuffs (Ie: Black will inflict withering; green will inflict poison, etc)

what about that?

**Comment by SteamDoge9000** [comment]
another event idea that can be exlcusive for this pack is when you have reached high warp levels that your machines (that require matemence) will have an event that will make them require matimence

**Comment by Prometheus0000** [comment]
About wither warp (discussed in discord), one way to give you some time to prepare is to have it first summon a 'ghost' version of the wither (with chat warning of impending doom or whatever), then summon the actual one a little later (10s?).

Other ideas: (I don't think I need to specify if they're temporary or not, do i?)
- Upside down screen
- Sideways screen
- Hide HUD
- Spawn different enemies, like a group of zombies outfitted with gear
- Spawn a companion cube from chisel
- Make it rain as a visual effect, but you can see that it's only around you (like 5 block radius)
- Drop fish instead of animals
- drop ocelots/wolves
- Change POV(?) like when slowed/picking up dolly, or the opposite
- Invert colors or similar
- Add a filter, like old-timey grain filter or black and white
- Start fires like holding firestone
- swap health and hunger values
- Change skin
- Invert direction controls
- Play some kind of sound, like enderman screech
- Other potion effects
- Super fast decay of blocks around you, only for a short time (creates a pit). Only chosen if you have enough vanilla materials (stuff it can affect) around you, since it would be useless otherwise

**Comment by SteamDoge9000** [comment]
some of those are cool but spawining fire around you as a warp event will get probematic real quick, becasue remember: Fire + machines = BOOM

Changing skins would be a nice one but how would that be pratical 

Super fast decay would be a pain in creating holes in your base as it would take time to fill them up and make the player (like I would) be parranoid about this effect happening

**Comment by Prometheus0000** [comment]
Maybe fire isn't the greatest idea. Or maybe a fake fire instead?

🤷 I'm just throwing stuff out. Last time I suggested spawning hungry nodes, I didn't expect anyone to like that idea either.

If we used the whole 200 range the giant hole one would be near the top. Also, if you chisel stuff, it no longer affects it, so only the cobble gang is going to be affected, and they deserve it.

**Comment by SteamDoge9000** [comment]
idk anything about that, but you fail to realize one thing. I belive most if not everyone uses cobble as a filler to fill most of the deep craters that can form and then top it off with the aproiate blocks (if they care about looks) other than dirt cobble is a comonly used resoruce for things and might cause mass damage to things that might be used to hold things up behind the scenes

also fake fire might not be a bad idea.

And the Ghost wither reminds me of that one thing that happens with Chance Cubes of a wither spawning and then vanishing when its "generation cycle" is complete (except for those times where its a real wither that will spawn and engage)

**Comment by D-Cysteine** [comment]
Thanks for all of the suggestions! Hmm, I was thinking we could lay out some more concrete goals / structure for this discussion. Apologies; this is going to be a rather long comment.

So are we all agreed that the wither should remain as the highest warp effect? If so, I propose we figure out what warp level we want to place it at, maybe set the warp level for the other effects in Warp Theory as well, and take a look at the gaps. Then, we can figure out roughly how many new warp effects we want to add, as well as their rough severity, and then we can look at all of the suggestions on this issue and pick out the ones that we like the most to implement.

I think there's already a few different attempts to assign new warp levels to the existing effects in this issue: https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/4270
so we could use those as a starting point.

---

For discussion of severity of warp effects, I was thinking that they fall into roughly four categories:

  1. **Mostly harmless** - effects that might be annoying, but will pretty much never cause harm to the player or loss of progress (by this I mean losing items, or damage to the player's base), regardless of when they occur. Examples would include fake explosion, bat swarm, friendly creeper.
  2. **Potentially dangerous** - effects that, on their own, are maybe minor, but could cause the player to die if they happen at a bad time. However, they will never cause loss of progress. Examples would include unnatural hunger, poison, anything that spawns hostile mobs.
  3. **Potential loss of progress** - effects that could potentially lead to loss of items or base damage, but can be avoided / fixed if the player is paying attention, or with proper base construction. Examples would include lightning (build a roof), thaumarhia (causes taint if not cleaned up), deadly gaze (can kill animals or villagers; prevent this by building walls).
  4. **Unavoidable loss of progress** - effects that are highly likely to lead to loss of items or base damage if they occur at a bad time, and cannot be avoided once they occur. The only way to avoid having these effects cause damage is to stay away from places where they can cause damage, which could mean having to stay out of your base entirely. Examples would include wind (due to breaking crops), wither.

I was thinking that we could place warp effects at warp levels according to which of these four categories they fall into, and we can see which categories have fewer warp effects and try to pick more warp effects to fill those in (except maybe the last category).

This ordering does lead to some potential weirdness in severity; like for example, a warp effect that just outright kills the player would be "less severe" than a warp effect that just places a redstone torch at the player's feet, because the former cannot cause loss of progress whereas the latter could lead to your LCR turning on and voiding all your input because the redstone control got messed up. But due to how expensive items and machines are in GTNH, I think that this ordering generally makes sense? There's definitely some room for discussion here though. Also, some things are just really annoying, and probably deserve to be placed higher because of that (random teleport, chest scramble).

Also, as SteamDoge pointed out, there's a few seemingly minor effects that could actually have huge consequences due to how GTNH works. Anything that creates fire could explode all of the player's GT machines, and anything that removes blocks could also cause this: like if the player just happened to use that block to contain the lava around their rock breaker, the lava could spill out and cause fires leading to an explosion. (If we do add effects like this, we should probably add a quest warning the player not to use such materials in their base.)

---

The last category, **unavoidable loss of progress**, is the one that I feel is most "unfun" right now. Since it's impractical to stay out of the base, the only ways to avoid these warp effects are to just avoid getting that much warp at all, or very careful planning and a far-away, sacrificial setup to quickly rush through all of the warp-inducing research and then eat a pure tear. I think that the main goal of this issue is to make this type of warp effect less unfun, by pushing these effects later in warp.

However, Prometheus's comment above about giving the player time to prepare brings up a good point. Another solution to these effects being unavoidable, would be to add in a delay time, giving the player time to get to a safe place out of the base. I think we could fairly easily do something like, when these effects trigger, we first send the player a chat message warning them that something bad is about to happen, and give them some time to get out of their base. Then we trigger the actual damaging warp event.

If we want to try this, I had some ideas:
 * We should add a quest somewhere where it will be seen by players accumulating high warp amounts. This quest could explain the delay mechanic, as well as tell the player what the warning messages / signs are.
 * The quest could also suggest (or give as rewards) getting a staff of traveling, and setting up a travel anchor in a warded jar / "panic room" somewhere safe. The staff of travel would allow the player to get to safety quickly regardless of where they are in the base. Players with extremely large bases would maybe need to build several panic rooms, which could be an interesting mechanic for base builders.
 * With the staff of travel technique, I feel like around 8 seconds of warning could be a good amount - short enough to be scary, but long enough to get to safety. Also, we'll probably want to add in audio cues for all of these, to make sure that players notice them.

**Comment by D-Cysteine** [comment]
Discussion with SteamDoge gave me an idea for a different way to potentially suppress warp effects while in base: armor that gives negative warping. So as long as you are wearing it, it'll reduce your effective warp. This should work even on servers where players can be logged off, as opposed to things like the soul brazier (and warp tumor? Not sure why that one is broken) which don't work as intended because players can be offline.

We can have the armor take damage over time or something, and so wearing it would be an ongoing cost to maintain it. Maybe have a version that uses EU instead, with a tiny battery, so that it would only work well in places where the player is able to set up a wireless charger to keep its charge up?

**Comment by SteamDoge9000** [comment]
While armor is not a bad idea, I do think it would be pointless unless it used an armor modification system like applying it to armors via a Infusion like that of runic shielding or to use a mix-n-match system like with HBM nuclear tech's armor modification system where it's simply like bauble slots for armor (certain modifications can only be applied to certain slots and armor [ie: a necklace like modifier can only be added to a chestplate slot, boot stuff exclusive to only footwear, etc] that allows to easily add and remove modifiers that exist between other amors (if you, for example, have modifiers that give the suit the ability to be treated as a hazmat that you added on an old armor, you can take them out and move them into the new armor)


Personally, I wouldn't go down this but like the idea of a multiblock machine that can be made to work either in protecting terrain from warp effects that cause damage (preferred) or it will act like a pylon made of bath salts and "asorb" (counter) the energy caused by the warp event into the machine when the player is within the machines radius of opperation (as some one helped me put it, like a wireless charger but for warp and does the opposite of what its supposed todo)

Due to the nature of warp and flux, the multiblock's "inner workings" can degrade as it handles this into biproducts (flux essentia and any other that may Be apropite as a runnoff); as a result, this multiblock will need to have matemence often

**Comment by bombcar** [comment]
Here's a warp effect - "Power has gone out from you" and all items in inventory/equipped have EU drained to zero.

**Comment by bombcar** [comment]
I like the idea of some form of grounding - even if only for certain warp effects or in certain cases.

Here's the thing - many players on servers with anti-grief (claims, etc) really don't have much of a downside for a wither besides free nether stars (and maybe killing some golems, etc). By the time you have enough warp to be causing withers, you should be adequately prepared to fight one.

Except for the base exploding part - for those without base protection, it's hugely dangerous. For those players something that can identify a warp effect and "cancel/ground" it at some cost could be nice, but this may be a separate thing/feature request.

I do like the idea/concept of "teching" magic more (also speaking of warp effects, some of the research says it pisses off witches, so witches should appear ...)

**Comment by SteamDoge9000** [comment]
Their also should be a messAge informing the player that a warp even has bee cancer if the machine does its job when within operating range

"*you feel as if a dark/twisted force was suddenly pulled away from your surroundings*"

**Comment by yukieiji** [comment]
One of the problems with Warp is that we don't know the amount of distortion that can be obtained through research. as a result, I think that there are many cases where the amount of Warp that produces a "potential loss of progress" effect is reached before one knows it.

Is it possible to let the players know the approximate amount of Warp will get when they finish the research?
This change should not require the implementation of items that prevent the Warp effect in the base, as we will be able to guess which stage of the Warp effect ours themselves are in, and use it to deduce whether the research will progress the stage or not.

**Comment by Prometheus0000** [comment]
> Is it possible to let the players know the approximate amount of Warp will get when they finish the research?

You already get told this in a lore-way, from 1-5 except the ones we change to be higher. People just don't care or pay attention.

**Comment by Prometheus0000** [comment]
> I do like the idea/concept of "teching" magic more

Personally I don't like this, but I don't like it in specific ways, like the EMT wand recharging thing that makes it seem like you just use tech to skip fundamental thaum stuff.

**Comment by SteamDoge9000** [comment]
> Personally I don't like this, but I don't like it in specific ways, like the EMT wand recharging thing that makes it seem like you just use tech to skip fundamental thaum stuff.

Well Magic and Tech have somewhat of a love-hate relationship from what I have observed in lore, though dispite this a connection has been made

The machine that I had described will need various forms of input, and the thing is theres one thing I thought of as well, this machine will be able todo it on not just one player, but can work wtih any player that has a warp event in its radius, though this will increase the consumption of resources but makes it so group research while working towards the pure tear or amulet can be easaly managed

**Comment by Prometheus0000** [comment]
I don't care about lore, what I care about is people ignoring the proper progression and just throwing EU or w/e at it so they can skip stuff. Also, rather than making a new thing, fixing the existing one seems better.

**Comment by Cata-bre** [comment]
> So are we all agreed that the wither should remain as the highest warp effect? If so, I propose we figure out what warp level we want to place it at, maybe set the warp level for the other effects in Warp Theory as well, and take a look at the gaps. Then, we can figure out roughly how many new warp effects we want to add, as well as their rough severity, and then we can look at 

I'd like to see the Wither at 200 Warp (per #5643) and warp effects scaled into your four categories. Gaps would be filled as needed.

1. Mostly harmless (0-50 warp)
2. Potentially dangerous (51-100 warp)
3. Potential loss of progress (101-150 warp)
4. Unavoidable loss of progress (151-200 warp)

Such scaling still forces you to deal with warp, but it allows for more exploration of Thaum + addons than is currently possible. Anyone in MV/HV/EV has to limit their warp until they can get the Pure Tear at IV. The game play experience of Thaum right now is balanced poorly w.r.t. warp; you have tons of consequences for researching at lower tech tiers, but once you hit IV all consequences vanish from crafting the pure tear.

Allowing players to research more while still dealing with effects makes the IV warp removal (Pure Tear) feel like less of a cliff.

This pack is the first time I've played with TC4, and I wasn't really sure what to research until I did some "meta" research on the Discord. It would be great if TC4 research felt less meta/had more options.

I agree with @Prometheus0000 that Thaum should be _insulated_ from EU usage (outside of what already exists in EMT).

**Comment by SteamDoge9000** [comment]
Well it docent have to be, EU, I was just suggresting EU as a power source for the machine to run, it can be something else, I just used EU because I wasnet aware of any other energy that the machine could use

It could run on essentia but im not sure if there are means for GT machines todo this

**Comment by Prometheus0000** [comment]
Why would we need a GT machine to do thaum stuff? The thaum mods exist, you know?

Not to mention that the topic is _new warp effects_.

**Comment by Cata-bre** [comment]
Is it possible to tie the base Thaumcraft warp effects to Warp Theory such that you can't afk in a hole until the warp effects disappear? If several of those, such as the Eldritch Guardian, could be tied to Warp Theory they'd be add thematic warp effects.

Ideas for warp effects (some of these cribbed from previous comments):
* Spawn Taintacles
* Spawn a zombie in Thaumcraft Armor
* Spawn a weak Crimson Portal that only spits out a few enemies
* Spawn Wisps
* Vis drain on a wand in inventory
* Wither potion effect
* Send the player to the Witchery dream world next time they sleep
* An "Unnatural Thirst" effect where you have to eat a blood item (Tinker's has several) to reduce the effect timer (very similar to Unnatural Hunger)
* Spawn tainted mobs (tainted chicken, tainted cow, etc.) OR convert a nearby friendly mob to its tainted version.
* Is it possible to generate fake mobs only visible to one player?
* Create faces in blocks that only appear for a second or so at a time. You look at a wall, see a face in the wall, and it disappears.

**Comment by Savallator** [comment]
Can we first please come to a conclusion about what we want to achieve? Because some of the effects here will just make players avoid TC as much as possible, as already a significant portion of the playerbase does.
Also, keep in mind you don't have the choice of what to research if you want to progress far in tc, because for unlocking Ichor you will absolutely need a significant portion of warp by researching items you otherwise don't even need.
I mean, if that is the challenge here i can also think of some even more annoying warp effects: 
-> Change a random stone in the vincinity, that is not touching air, to a spawner of type Wither
-> Scramble items in any nearby inventories around (funny with nukes!)
-> drain nearby Battery buffers of power
-> Randomly disconnect/connect nearby pipes and cables
-> have the player emit redstone signal to its environment for a short time

But honestly, what do you want to achieve with warp? Because in it's current state, stuff like the Decay/Life effect is just destroying bases, the lightning is annoying every player on the Server and also very buggy, and some of you want even more annoying effects. Do you really want the playstyle to be "don't do any warp stuff until you are ready, then get a seperate base, do everything warp related and quickly reset warp"? Because that will actively drive people away from the pack.

**Comment by D-Cysteine** [comment]
> Can we first please come to a conclusion about what we want to achieve?

Agreed; I think that this is the most important question, which we need to answer first. I've been busy with other things lately, but once I have more time, I was planning on trying to direct the conversation here back to answering this question.

I think a good starting point for this is figuring out roughly what effects should happen at which warp levels, like in [this comment](https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/8611#issuecomment-931407879) or [this comment](https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/5643#issuecomment-591080795).

**Comment by Cata-bre** [comment]
> Can we first please come to a conclusion about what we want to achieve?

I want TC gameplay to be less all or nothing. @Savallator describes the problem with the current playstyle well:

> Don't do any warp stuff until you are ready, then get a separate base, do everything warp related, and quickly reset warp

I want to achieve a better warp/gameplay curve for Thaumcraft + addons that will:

* Cause more players to play with Thaumcraft
* Allow for a better risk/reward ratio
* Still be punishing at high levels of warp

Currently the risk (warp effects) do not scale well with the rewards (multiple Dangerous/Taboo researches). Many players wait until IV/pure tear to completely circumvent this risk-reward curve. This means they aren't engaging with Thaumcraft or warp at all.

By increasing the Warp Theory range we can make the Thaumcraft risk/reward curve more engaging so players don't circumvent it. We also give players more options/choices/research paths. Want to heavily research Blood Magic, but not other pages on the Thaumonomicon? That's viable. If you want lots of Blood Magic research and other tabs on the Thaumonomicon? Well, you'll pay for that in warp effects. Completely unlocking multiple pages will still have consequences. I'll make a separate comment with a table detailing (what I think) is a better warp risk/reward curve.

**Comment by Savallator** [comment]
It destroys the blocks behind the wall, and under the floor. **That** is my problem. That is why i suggested it only affecting blocks you are looking at, so only visible blocks. I still find patches of sand and gravel way deep down from the few times i had the effect back then before i disabled it.

**Comment by Savallator** [comment]
That is the point - if you make warp _too_ annoying, people will just not play TC because of it, like it already happens for a major part of the playerbase. Also, mechanics like the ichor unlock and stupid research requirements force you to get a large amount of warp to progress to the good stuff. I would rather make warp effects more harmless, let the total warp level decay over time (or let the research give only temp warp), but make actually crafting/using tc stuff give more, and also remove warp removal items. that way thaumcraft moves away from "either don't use it so it does not ruin your base, or rush through it at IV in a seperate base and evade all the effects". Since i play on a private server, i have no problems with warp, i will just disable the effects that are destroying the fun. But on public servers you see most of the people staying away from thaum and missing out on a big amount of content just because some people find it fun to have an annoying mechanic punishing players for progressing.

**Comment by SteamDoge9000** [comment]
agreed
if that is to be eft as is or made to work on chiseled blocks then it should be reserved for higher level warp effects
like close to the wtiher new rate or something

**Comment by DisasterMoo** [comment]
Might add to this: I've recently crafted a 10k primordial aspects node matrix. Got about 30k normal warp and ate a pure tear. About 3hrs of warp effects, which didn't bother much but what was really annoying were the teleportation and thunder warp. It kept occurring for about 3 days (left afk) making the game impossible to play until it went off.

Some suggestions:
- Pure tear warping effects fixed (ie: 500 random effects regardless of warp amount)
- Teleporting warp only sums up to a fixed amount / reset each time it is added
- Pure tear doesn't cause teleportation warp.

**Comment by Prometheus0000** [comment]
You can just die after eating it and it'll reset.

**Comment by DisasterMoo** [comment]
I tried dying after it was taking half a day, didn't really reset the warping. Guess you have to do it just after eating the pure tear them. But anyway, hardcore players (no death run), or, if you wanted the nether stars.

**Comment by SteamDoge9000** [comment]
maybe intstead of doign it like that, a more direct way of realeaseing the energy shoudl be done

if the user is at the certain point with total warp and eats the tear, it may do something more drastic, as it will not only spawn withers everywhere or something, but it may also cause a massive change to the land, 

maybe a massive release results in the land become tainted and most blocks within a large radius become transformed into flux waste and what not

this can make a artifical taint biome if it gets to let say, 500 or more warp is present when you ear the tear and the range of this transmutation increases for each level higher

this would be a great drawback but better than crapping wtihers out. and lets be honnest, if they use that at their base it might be game over for them but if not it will still be somewhat of a massive cleanup project


or it could make a large amount of the land become erie

**Comment by bombcar** [comment]
Almost certainly thermos as that’s our server

**Comment by D-Cysteine** [comment]
Alright, sorry for the long delay, but the new warp effects are finally done / in a PR! This is going to be another super long post, as I added a bunch of stuff. Some of these are from various ideas above (thanks for the suggestions!) and some I came up with as I was working on this.

Also, I noticed while working on this that 200 is really just an arbitrary limit in the code. I'm pretty sure that we can easily change this maximum to whatever we want. So that's something to keep in mind while balancing this.

---

First, here's my new proposed table of warp effects, which is based on [this comment](https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/8611#issuecomment-936370421):
| **Warp Effect** | **Warp Amount** |
| ------------- | ------------- |
| **Mostly harmless** | **0-50** |
| *Eye Blink* | *5* |
| Fake Explosion | 10 |
| *Coin* | *15* |
| *Pumpkin* | *15* |
| *Phantoms* | *20* |
| *Lay Eggs* | *20* |
| Fake Creeper | 25 |
| *Fake Rain* | 25 |
| Spawn Bats | 30 |
| Spawn Friendly Creeper | 30 |
| *Fake Enderman* | *35* |
| Nausea | 40 ​|
| Jump Boost | 40 ​|
| Blood | 45 ​|
| *Vis Drain* | *50* |
| *Insomnia* | *50* |
| **Potentially dangerous** | **51-100** |
| Thunderstorm | 55 |
| Blindness | 60 |
| Poison | 60 |
| *Inventory Swap* | *65* |
| *Junk* | *70* |
| Livestock Rain | 70 |
| *Spawn Fire Bats* | *75* |
| *Spawn Doppelgänger* | *75* |
| *Wither Potion* | *80* |
| *Spawn Angry Endermen* | *80* |
| Wind | 85 ​|
| *Ender Pearl* | *90* |
| *Blaze Fireball* | *95* |
| *Litmus Paper* | *100* |
| *Obsidian* | *100* |
| **Potential loss of progress** | **101-150** |
| Lightning | 110 ​|
| Random Teleport | 120 ​|
| *Mushrooms* | *130* |
| Acceleration | 140 |
| *Countdown Bomb* | *150* |
| *Fake Wither* | 150 |
| **Unavoidable loss of progress** | **151-200** |
| *Snow* | *160* |
| Decay | 170 |
| Random Trees | 180 |
| Chest Scramble | 190 |
| Spawn Wither | 200 |

I've omitted all base Thaumcraft effects, because they fade away over time, and I'm not sure if we can modify their warp levels due to licensing issues. I was particularly unsure about the events around 51-100 (**Potentially dangerous**); would appreciate help with making sure that those are in a good order.

---

Okay, now for the changes to existing warp events, and new warp events. First, I fixed an annoyance with Random Teleport: it now reduces the remaining # of teleports by 5% of the current value. This prevents being stuck with random teleport for hours after doing something like eating a pure tear.

I added "multi-level" warp effects. Multi-level warp effects can have stronger effects, if you have a lot of warp. The way that it works is, let's say that we have a warp effect that triggers at 10 warp. If you have 10-20 warp, you'll only get the base effect. If you have 21-30 warp, you have an increasing chance of getting level 2 of the effect, maxing out at 50% at 30+ warp. If you have 31-40 warp, you have an increasing chance of getting level 3 of the effect, maxing out at 25% at 40+. The idea is for lower-level warp effects to stay interesting at higher warp / after the player is used to them, by having a "twist"; the higher-level effects aren't intended to behave very differently from the base effect.

I added multi-level effects for these old warp events:
 * Spawn Bats (30 warp)
   * Level 2 (>60 warp): bats are on fire
   * Level 3 (>90 warp): bats are on fire, and have fire resistance
 * Spawn Friendly Creeper (30 warp)
   * Level 2 (>60 warp): "friendly" creeper will explode (doing no damage)
   * Level 3 (>90 warp): same as level 2, except the creeper is invisible until right before it explodes
 * Blood (45 warp)
   * Level 2 (>90 warp): blood is black
 * Livestock Rain (70 warp)
   * Level 2 (>140 warp): rains squid
   * Level 3 (>210 warp): rains tainted livestock. Note that it's possible for some of the livestock to survive if there's a ledge or something nearby, and tainted sheep can spread taint, so this one is a bit dangerous.

Now for the new effects:
 * Eye Blink (5 warp): player blinks a few times (Blindness for 1 second). Note that this will drain durability from the Talisman of Remedium.
 * Coin (15 warp): player gets a Thaumcraft coin
 * Pumpkin (15 warp): if player has no helmet, places a pumpkin in the helmet slot
 * Phantoms (20 warp): spawns some mobs which do nothing, and vanish when the player looks at them
 * Lay Eggs (20 warp): player lays a few eggs at random intervals
 * Fake Rain (25 warp): fake rain particles spawn around the player. Visible only to the player.
 * Fake Enderman (35 warp): plays a distorted enderman scream sound for the player
 * Vis Drain (50 warp): drains a bit of vis from player's wands
 * Insomnia (50 warp): player cannot sleep for a few minutes
 * Inventory Swap (65 warp): does one of the following at random:
   * If player is holding an item and has an empty slot in inventory, places the item in hand into the empty slot
   * Randomly swaps the items in two slots in inventory (not hotbar)
 * Junk (70 warp): fills all empty inventory or hotbar slots with random junk items
 * Spawn Fire Bats (75 warp): spawns some Thaumcraft fire bats. Multi-level:
   * Level 2 (>150 warp): fire bats can randomly be explosive or devil fire bats (more info: [Nine Hells upgrades](https://thaumcraft-4.fandom.com/wiki/Focal_Manipulator))
 * Spawn Doppelgänger (75 warp): spawns a mob that does the following:
   * Follows players around just like the friendly creeper
   * Looks like the player that spawned it, and has the same max and current health
   * Heals over time, about as fast as a well-fed player in vanilla Minecraft
   * Whenever it takes damage from any source, deals that same damage to the player that spawned it, ignoring armor
 * Wither Potion (80 warp): gives the player 30 seconds of Wither 3, Hunger 3, Slowness 3
 * Spawn Angry Endermen (80 warp): only triggers at night. Spawns a few endermen who will immediately attack the player. Multi-level:
   * Level 2 (>160 warp): endermen have Invisibility and Strength 2
 * Ender Pearl (90 warp): makes the player fire an ender pearl wherever they happen to be looking
 * Blaze Fireball (95 warp): after a 2-second delay, player fires three blaze fireballs. There's a delay so you have time to look away from anything flammable.
 * Litmus Paper (100 warp): gives the player some litmus paper. This one is meant to warn the player that they are getting dangerous amounts of warp. Can only trigger once per 8 hours.
 * Obsidian (100 warp): spawns a block of obsidian where the player's head is, suffocating them
 * Mushrooms (130 warp): randomly places mushrooms, converts grass and dirt to mycelium, and converts cows to mooshrooms
 * Countdown Bomb (150 warp): player gets a countdown from 10 seconds. Once the countdown hits 0, triggers a real creeper explosion centered on the player. Also spews particles everywhere so other players know to run away.
 * Fake Wither (150 warp): plays a wither spawn sound effect. Like the litmus paper, this is meant to warn the player that they are getting dangerous amounts of warp.
 * Snow (160 warp): randomly places snow, converts water to ice, and spawns snow golems with fire resistance. The snow golems can quickly make a massive mess, which is why I placed this one so high.

**Comment by bombcar** [comment]
This looks really good - as for livestock rain - I've been getting that "indoors and underground" - so something's a bit off there (they land on the roof and survive? die? I don't know, but I find wool days later).

**Comment by D-Cysteine** [comment]
The way that livestock rain works is, it picks a random block within a 15x15x15 cube centered on the player (closer to player is weighted more heavily). Then, if that block and the 25 blocks above it are all air, it spawns livestock at the top of that 25-block column.

So even if you are under a roof, if you are within 7 height of the surface, there's a chance that it'll pick a random block on the surface and be able to spawn. Or if your rooms are taller than 25 blocks, it could even spawn indoors.

**Comment by bombcar** [comment]
That explains it - though note I got it on the moon once so "air" is loosely defined. They *SHOULD* have oxygen setups in that case I guess ;)

**Comment by bombcar** [comment]
As to the comment in Discord about the chest scramble only affecting wooden chests - I propose the following:

1. Wooden only (literally single and doubles only) be moved lower, but the text imply that metal chests may help ....
2. Wooden - iron be about where it is and a noise suggesting "precious metals might help"
3. All "normal" chests wooden -> diamond with a "crystal may help"
4. Optional: scramble all chests and disconnect ae2 drives (!!!)

**Comment by D-Cysteine** [comment]
Hmm, the chest scramble is kind of tricky. So as far as I'm aware (and someone please correct me if I'm wrong), there is no good way to check if something is a "chest". We can check if something is an "inventory", but that would include a lot of stuff that we probably don't want to include, such as machines and maybe even item pipes.

So we'd have to hard-code a list of chests, but there's a few issues with that:
 * Need to have Warp Theory depend on all of these other mods that add chests
 * Have to keep this updated whenever we add or remove mods
 * If we miss any, it'll feel really weird. Why are my diamond chests being scrambled, but not my skystone chests? Etc.

Given this, the fact that late-game AE setups mostly just avoid this entirely, and the fact that especially in a tech-heavy modpack like GTNH, scrambling chests just feels really bad, I propose that we disable the chest scramble effect entirely. We could shift things around (maybe move fake wither sound effect to 190), or we could just leave 190 empty and have a 20-warp gap before the wither effect? I could also add a new effect for 190.

**Comment by bombcar** [comment]
We could use all the "Iron Chest" mod chests - but I agree.

What about "scramble player inventory + backpacks" ?

Also maybe Fake Explosion could move up a bit or change it so the first "fake explosion" is the "steam machine out of power" noise, the second one a few warp levels later is "electric machine out of power noise" (may be the same noise?) and the third highest is "electric machine rapidly disassembling itself noise).

I know the explosions I just ignore, but power out noise wakes me up.

**Comment by D-Cysteine** [comment]
Yeah, agreed, if we wanted to keep Chest Scramble and make it apply to more things, Iron Chests would probably get us most of the way there. The other popular ones that I was thinking of were the Avaritia compressed chest, and maybe the AE2 skystone chests?

Scramble backpacks might have the same problem: there's a fair number of backpack mods. I guess just scrambling the Backpack mod's backpacks will get most of them, but there's also:
 * Adventure backpack
 * IC2 backpacks
 * Witching Gadgets bag
 * Extra utilities might have a bag of holding?

Just scrambling the player's main inventory is really easy though; there's even already an effect that does a weaker version of it. That could be a good replacement for Chest Scramble? Maybe it could go at 150, and push Mushrooms and everything after that down by 10?

The GT noises is also a great idea. Probably best to add it as its own separate effect; makes it easier to disable when GT isn't present. Maybe at 35? Something like:
 * Level 1 (35 warp): vent steam
 * Level 2 (70 warp): machine out of power
 * Level 3 (105 warp): I see something in the sound files called "NukeExplosion", so how about that one

I'll work on adding these when I have more time, hopefully later today.

**Comment by SteamDoge9000** [comment]
Here's two

Semi-dangerous

"You feel as if your Armor was made of gelitan..."

The affect: armor stats reduce greatly or any armor points you have will be nullified (armor = 0 regardless of what you wear?)

The only way to possible way to reverse this effect is to "bake" yourself in fire

"You feel your armor has become solid once more..."

Warp effect will only trigger if you have armor on or it will happen regardless (you will take more damage if you wear no armor)

Dangerous

"You suddenly feel as if your body is under a great amount of stress and more fragile than usual"

The affect: you body becomes like Tempered Glass! Attacks deal 2-3X more damage. Critical hits have a Chance to make you shatter (Inatant death)

Projectiles can have a chance to one shot you, EXPLOSIONS ARE ALWAYS FATAL REGARDLESS of what they are or what your wearing (unless it's some high end stuff like the Draconic armor sets)

Falls between can cause to to take more damage or "crack" (being cracked significantly increases the chances of breaking). Falls 5 blocks or higher will kill you instantly (cus dropping glass from 5 or more meters is guaranteed to break

Remedies: Defence potions will produce the message "you feel as if your body is less brittle" and the damange multiplayer is reduced to 1.5 (regardless of what level of potion power it is)

Potions of strength will restore your body composition that was altered from this warp event (will cancel the effect line eating flesh or brains reduces unnatural hunger)
"Your body no longer feel as fragile as glass


Duration will depend on how much total warp you have accumulated

**Comment by Cata-bre** [comment]
> Just scrambling the player's main inventory is really easy though; there's even already an effect that does a weaker version of it. That could be a good replacement for Chest Scramble? Maybe it could go at 150, and push Mushrooms and everything after that down by 10?

I would put chest scramble and player inventory scramble up fairly high. As well as the hotbar to inventory scramble. Other than that, I like the table of warp effects you've constructed. I think it will make Thaumcraft more approachable, yet still punishing at high levels.

**Comment by Cata-bre** [comment]
I think 200 warp as the upper limit is a good starting place for now. It increases our warp range significantly, but is low enough that you're still limited in the amount you can research without serious consequences.

**Comment by D-Cysteine** [comment]
Thanks again for the suggestions! I made the two changes mentioned here:
 * https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/8611#issuecomment-958219440

and removed Chest Scramble and updated the table:
 * https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/8611#issuecomment-958003423

I think the list of effects is looking pretty good now; filled in some of the gaps.

---

Here's what Junk spawns currently:
 * https://github.com/GTNewHorizons/WarpTheory/blob/44a96b42c4500b188e4b75762fb17ba991c9f5ee/src/main/java/shukaro/warptheory/handlers/warpevents/WarpJunk.java#L16-L33

I can't decide if it's better or worse than Inventory Scramble. I have it after Inventory Scramble right now, but maybe we could move it earlier? All three of the inventory effects are pretty high right now purely due to how annoying they can be. I made sure never to place items into the currently selected slot so you won't do something like accidentally mallet a transformer and blow up your base.

**Comment by Cata-bre** [comment]
> I can't decide if it's better or worse than Inventory Swap.

While annoying, I would find the Junk effect less frustrating than inventory swap. At high levels of warp (200+) you could add a Level 2 to the Junk effect that spawns Taint Tendrils and Tainted Goo. This is as dangerous as Livestock Rain spawning tainted sheep.

> I made sure never to place inventory items into the currently selected slot so you won't do something like accidentally mallet a transformer and blow up your base.

👍

**Comment by D-Cysteine** [comment]
Ok, I swapped Inventory Swap and Junk.

I *think* the Taint Tendrils and Tainted Goo aren't actually capable of spreading taint? They can just decay in your inventory, which gives you a damaging debuff IIRC.

**Comment by Cata-bre** [comment]
I thought the Taint Tendrils and Tainted Goo gave you temporary warp. You're correct, they aren't near as dangerous as tainted sheep.

**Comment by Cata-bre** [comment]
Just to be clear, the updated table values are now in the 2.1.20, correct?

**Comment by bombcar** [comment]
2.1.2.0 doesn't exist yet, but they're in 2.1.1.5pre2 if you want to take a look in single player: http://downloads.gtnewhorizons.com/Dev-Pack/Pack/Client/

**Comment by meonester** [comment]
I'm currently playing on version pre.2.1.1.5-dev2. Those new Warp "prizes" needs some tweaking as they are literally getting spammed (1 per minute or so). For instance flux flu almost never goes away, and during last hour my character laid ~20 eggs. That said i have only 21 warp(17 permanent + 4 normal).

**Comment by bombcar** [comment]
Perhaps we need an inherent brake on the amount per X minutes that loosens up as you go higher?

**Comment by D-Cysteine** [comment]
Flux flu is a vanilla Thaumcraft warp effect, so it's unrelated to these changes. Vanilla Thaumcraft warp effects do fade over time, so as long as you aren't gaining any more warp (not sure if removing warp also refreshes them), they should gradually stop affecting you at all. This might explain why you noticed things stabilizing: it could just be the vanilla effects fading away.

The egg effect queues up multiple eggs each time it triggers (3-8 eggs); it's not one egg per trigger. So those 20 eggs might actually only be 3 events. Also, with such a low warp, you are only eligible for a short list of warp events. This means that you're much more likely to get the same one repeatedly (i.e. eggs multiple times), since it only picks from among events that you're eligible for.
